proposals/2023/03-15_init-step.md

```diff
@@ -0,0 +1,341 @@
+# Add `InitStep` resource
+
+<!--
+The name of this markdown file should:
+
+1. Short and contain no more then 30 characters
+
+2. Contain the date of submission in MM-DD format
+
+3. Clearly state what the proposal is being submitted for
+-->
+
+| Key           | Value              |
+| :-----------: | :----------------: |
+| **Author(s)** | Jacob Floyd        |
+| **Reviewers** |                    |
+| **Date**      | March 15th, 2023   |
+| **Status**    | Waiting for Review |
+
+<!--
+If you're already working with someone, please add them to the proper author/reviewer category.
+
+If not, please leave the reviewer category empty and someone from the Vela team will assign it to themself.
+
+Here is a brief explanation of the different proposal statuses:
+
+1. Reviewed: The proposal is currently under review or has been reviewed.
+
+2. Accepted: The proposal has been accepted and is ready for implementation.
+
+3. In Progress: An accepted proposal is being implemented by actual work.
+
+NOTE: The design is subject to change during this phase.
+
+4. Cancelled: While or before implementation the proposal was cancelled.
+
+NOTE: This can happen for a multitude of reasons.
+
+5. Complete: This feature/change is implemented.
+-->
+
+## Background
+
+<!--
+This section is intended to describe the new feature, redesign or refactor.
+-->
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+<!--
+Provide your description here.
+-->
+
+Replace the injected "init" step (a pseudo-container) with `InitStep`, a new resource.
+
+Each `Build` will have one or more `InitSteps` that logically group related parts of the log data.
+
+An `InitStep` is a report by a "reporter" about a discrete part of the build setup. The "reporter" would be a logical part of the stack, like the "Pipeline Compiler" or the Runtime. For example, we could have these reporters be discrete initsteps (Reporter: name of the `InitStep`):
+- Pipeline Compiler: report info/debug logs
+- Docker Runtime: report network setup
+- Docker Runtime: report volume setup
+- Kubernetes Runtime: Pod YAML
+
+`InitStep`s are always part of a build. An `InitStep` may also be associated with a `Step` or a `Service`. Today, the runtime prefixes each step's log with the command (`$ command ...`). An `InitStep` linked to a `Step`--or a `Service`--will allow the worker to cleanly report what it runs separate from the output of that command. The UI, in turn will also be able to visually separate the command and output. The worker could also log other information like environment vars, or for kubernetes, the YAML container definition in the pod.
```

Comment:
So we'll have the `Log` for the `InitStep` contain the `$ command ...` portion for each `step`/`service` from:

https://github.com/go-vela/server/blob/main/compiler/native/script.go#L70-L98

And then the `Log` entry for that `step`/`service` will contain the output from invoking that command?

Reply:
Yes. I want to separate the Vela generated bits from the command step or service output.